src/main/java/org/sagebionetworks/bridge/hibernate/HibernateSurveyDao.java

```diff
@@ -0,0 +1,365 @@
+                return results.get(0);
+            }
+        }
+        
+        return null;
+    }
+
+    @Override
+    public List<Survey> getAllSurveysMostRecentlyPublishedVersion(StudyIdentifier studyIdentifier,
+            boolean includeDeleted) {
+        ImmutableMap<String, Object> params = ImmutableMap.of(
+                "studyKey", studyIdentifier.getIdentifier());
+        
+//        String countQuery = SELECT_COUNT + GET_ACTIVE;
+        String getQuery = SELECT_TEMPLATE + GET_ALL_ACTIVE_PUBLISH;
+        
+        List<HibernateSurvey> results = hibernateHelper.queryGet(
+                getQuery, params, OFFSET, 50, HibernateSurvey.class);
+        
+        return findMostRecentVersions(results);
+    }
+
+    @Override
+    public List<Survey> getAllSurveysMostRecentVersion(StudyIdentifier studyIdentifier, boolean includeDeleted) {
+        ImmutableMap<String, Object> params = ImmutableMap.of(
+                "studyKey", studyIdentifier.getIdentifier());
+        
+        String getQuery = SELECT_TEMPLATE + GET_ALL_SURVEYS;
+        
+        List<HibernateSurvey> results = hibernateHelper.queryGet(
+                getQuery, params, OFFSET, 50, HibernateSurvey.class);
+        
+        return findMostRecentVersions(results);
+    }
+    
+    /**
+     * This scan gets expensive when there are many revisions. We don't know the set of unique GUIDs, so 
+     * we also have to iterate over everything. 
+     * @param surveys
+     * @return
+     */
+    private List<Survey> findMostRecentVersions(List<HibernateSurvey> surveys) {
+        if (surveys.isEmpty()) {
+            return ImmutableList.copyOf(surveys);
+        }
+        Map<String, Survey> map = Maps.newLinkedHashMap();
+        for (Survey survey : surveys) {
+            Survey stored = map.get(survey.getGuid());
+            if (stored == null || survey.getCreatedOn() > stored.getCreatedOn()) {
+                map.put(survey.getGuid(), survey);
+            }
+        }
+        return ImmutableList.copyOf(map.values());
+    }
+    
+    String generateGuid() {
+        return BridgeUtils.generateGuid();
+    }
+    
+    // Package-scoped for unit tests.
+    void deleteAllElements(String surveyGuid, long createdOn) {
+        ImmutableMap<String,Object> params = ImmutableMap.of(
+                "surveyGuid", surveyGuid, 
+                "createdOn", createdOn);
+        
+        hibernateHelper.query(DELETE_SURVEY_ELEMENTS, params);
+    }
+    
+    private void attachSurveyElements(Survey survey) {
+        ImmutableMap<String, Object> elementParams = ImmutableMap.of(
+                "surveyGuid", survey.getGuid(),
+                "createdOn", survey.getCreatedOn());
+        
+        String getElementQuery = SELECT_ELEMENT_TEMPLATE + GET_ELEMENTS;
+        
+        List<HibernateSurveyElement> elementResults = hibernateHelper.queryGet(
+                getElementQuery, elementParams, OFFSET, 50, HibernateSurveyElement.class);
```

Comment:
In the off-chance there are more than 50 survey elements, we want to return all of them.